### DISCUSSION TITLE
GSOC 2023: MPM Project ideas

### DISCUSSION THREAD

**Discussion by kks32** [cb-geo/mpm#740] [Ideas]
# Project Ideas
## What is CB-Geo MPM
CB-Geo Material Point Method (MPM) is a large-deformation code for modeling particulate and fluid systems. MPM is a hybrid method that employs particles that traverse a background mesh to represent the physical state of the system.

##  Project 1: Differentiable Programming in MPM

Differentiable programming combines automatic differentiation with gradient based optimization that allows for solving inverse and design problems. Differentiable programming is a method of quickly solving complex numerical functions and combines two components: (1) automatic differentiation and (2) gradient-based optimization. In recent years, this method has gained popularity as a machine learning tool in the scientific community, with many applications including differentiable ray tracing algorithms or probabilistic programming. 

Solving optimization and inverse problems are critical for engineering design and analysis. Despite the computational power, calculating derivatives, i.e., evaluating the forward simulation multiple times with small perturbations, is computationally very expensive and is prone to numerical instability. Traditional forward simulations cannot be used in machine learning models for optimization, as they cannot compute gradients with respect to the input parameters (reverse mode). This project involves developing a novel Differentiable Programming simulator that combines automatic reverse differentiation with a second-order gradient-based optimization algorithm, such as L-BFGS, to develop a fully-differentiable MPM simulator. Using the differentiable simulator, we can identify the input material properties by iteratively updating the input parameters by minimizing a loss function. Typically the loss function is the norm of the difference to the target observation. We then minimize this gradient (loss corresponding to the input parameters) using a second-order optimization algorithm. Differentiable MPM solves the exact PDE and can approximate the dynamics of its environment better than model-agnostic reinforcement learning methods. The differentiable MPM simulator will be written in Google JAX and exploits Tensor Processing Units to accelerate the computational time. We will demonstrate the functioning of the diff MPM code by extracting the input parameters for a collision of two deformable bodies. The differentiable MPM is a novel tool to provide gradients through a simulator, which can be coupled with existing machine learning algorithms to generate real-time decisions and optimization in robotics.

| **Intensity**                          | **Priority**              | **Involves**  | **Mentors**              |
| -------------                          | ------------              | ------------- | -----------              |
| High  350 hrs |  High  | Integrating differentiable programming in CB-Geo MPM for optimization, design and solution to inverse problems. | [Krishna Kumar](https://github.com/kks32)  | 

The main goals of this project are to:
- Develop a differentiable MPM code in 1D, 2D and 3D. A sample 1D differentiable MPM code is available at [1D DiffMPM](https://github.com/geoelements/diff-mpm/blob/main/diffmpm-1dbar.py)
- Develop example benchmarks for differentiable MPM.
- Solve an inverse problem to identify input parameters given a runout.

Skills required:
- Python and C++
- Experience with JAX or PyTorch differentiable frameworks. 

Expected outcomes:
A prototype that offers the features of differentiable programming to MPM to solve inverse, optimization, and design problems. We'll solve for input parameters given an output result variable using different optimization routines like ADAM and SGD. 

Skills required:
- Python and C++
- Experience with JAX or PyTorch differentiable frameworks. 

## Benefits of working on this project
Students who work on this project can expect their skill-set to grow in
- AI/ML applied to scientific computing
- Accelerating numerical simulations with AI/ML 
- Scientific Machine Learning

## Benefits to project/community
- Ability to solve optimization and inverse problem
- Couple ML with simulations to enable data-driven discoveries. 

## Helpful Experience
- Have a working knowledge of Python/JAX
- Experience with writing machine learning applications in JAX, TF or PyTorch

## First steps
- Become familiar with current simulation and visualization infrastructure in CB-Geo MPM.
- Write a simple 1D DiffMPM with multiple material points per element [1d MPM](https://www.geoelements.org/LearnMPM/axial-vibration1d.html) using JAX.
- Discuss with Mentors on GitHub

**Comment by SachinJalan** [comment]
Hello Sir,
I am interested in this project. I have gone through Wikipedia page on MPM and the theoretical manual of CB-Geo MPM, can you please provide some more resources to aid my understanding of the topic which helps me in completing the first steps of making 1-D DiffMPM.
Thanks.

**Reply by kks32** [reply]
Hi @SachinJalan Thanks for reading through. Please have a look at https://www.geoelements.org/LearnMPM/point-axial-vibration1d.html#d-mpm-code-complete. Could you please write this in JAX?

**Reply by SachinJalan** [reply]
Hello @kks32 ,
I have implemented the given code in JAX can you please comment on further improvements. I have used Just In Time compilation for updating the values per iteration and used JAX arrays.
[implementation](https://github.com/SachinJalan/MPM_1D_Jax/blob/main/MPM_1D_UsingJAX.ipynb)
Thanks.

**Reply by kks32** [reply]
Hi @SachinJalan That's great. Could we have a chat: https://calendly.com/krishnakut/30min book a time that's convenient for you. We can also find a different time if this does not work for you.

**Comment by dhanush-2501** [comment]
@kks32 HI sir, I am interested in this project and I went through the resources you provided , can I start working on it ?
